### PR TITLE
PP-315: Fix for deleting subjob in routing queue

### PR DESCRIPTION
**Pull request by dilip-krishnan**
#### Issue-ID

* **[PP-315](https://pbspro.atlassian.net/browse/PP-315)**

#### Problem description
* Deleting a subjob of a Job Array in User H state in Routing queue causes jobs to fail to route

#### Cause / Analysis
* Whenever state of a subjob changes to any other state other than queued, the parent array job is set to JOB_STATE_BEGUN. This state(JOB_STATE_BEGUN) of parent job was not handled in job_route(), due to which server fails routing the job.

#### Solution description
*  To allow parent job with JOB_STATE_BEGUN to route.
 
#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

Issue: Deleting a subjob of a Job Array in User H state in Routing queue
causes jobs to fail to route

### CODE REVIEW

**Comment by ashwathraop** [comment]
Code looks fine to me. I sign off.

File: test/tests/pbs_job_array_in_routingqueue.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+                                  ATTR_h: 'u',
+                                  ATTR_J: '1-2:1',
+                                  ATTR_r: 'y'})
+
+        jid = self.server.submit(j)
+        self.server.expect(JOB, {ATTR_state: 'H'}, jid)
+        self.server.expect(JOB, {ATTR_state + '=Q': 2}, count=True,
+                           id=jid, extend='t')
+        subjobs = self.server.status(JOB, id=jid, extend='t')
+        # Delete one of the subjob
```
Comment (agrawalravi90) [inline test/tests/pbs_job_array_in_routingqueue.py]: Can you expand this comment to explain what you are expecting to happen when the subjob is deleted?

**Comment by agrawalravi90** [comment]
"Whenever state of a subjob changes to any other state other than queued, the parent array job is set to JOB_STATE_BEGUN"

One dumb question: does this mean that if a sub-job is deleted while the job array is in held state (so its state changes to 'X'), it will change the state of the parent job to begun? is that the expected behavior? if not, should we fix the code so that this transition doesn't happen?

**Comment by dilip-krishnan** [comment]
Hi Ravi, 
       Yes, the state change of Parent job to begin is correct.  I can understand the confusion since begin sounds like the job is actually running. But that's not what "Begin" is here. In context of array job "Begin" means any kind of operation done on the subjobs of an array job. this could mean run, delete, release,
re-queue

**Comment by agrawalravi90** [comment]
Thanks for clarifying Dilip. However, I'm still not sure about this. The admin guide (table 6-20) says that the Begun state means that the array job has "started". In the code (svr_evaljobstate()), we set the parent job's state as Begun if all the subjobs aren't queued, which is why this probably happens, but it hasn't been explained why. The other place where we are setting it to Begun (complete_running()) there's a comment saying:
		/* if this is first subjob to run, mark */
		/* parent Array as state "Begun"	*/
So here the association is with running a sub-job.
Is there any documentation which says that we should change the state of parent job to Begun if a subjob is deleted? If the job is in Held state, we are changing it to Begun state after deleting a subjob, that just sounds odd to me. If this is indeed the expected behavior then we should at least document this somewhere for the users as this is an external interface.

**Comment by billnitzberg** [comment]
The original intention of "Begun" is that "subjobs have started running".  I think we have some flexibility (as I don't believe we define it exactly in the docs), but if we want to be nice to our customers, we should not mark a Job Array as "Begun" if no subjoins have ever been been started.

**Comment by agrawalravi90** [comment]
Thanks for the input Bill! 

There's also the fact that we do not want to route jobs in the held state unless the 'route_held_jobs' flag is set (admin guide section 2.2.6.6), so by changing the state to "Begun" from "Held" we might end up routing jobs which were not supposed to be routed?

@dilip-krishnan : if it's not documented that we should change the state to "Begun" from "Held" when a subjob is deleted, then I feel like we need to fix the code such that this transition does not happen when a subjob is deleted. What do you think?

**Comment by dilip-krishnan** [comment]
Hi Bill & Ravi,
I had the same impression about this, when i was first fixing it for earlier release. But after discussing it with @subhasisb , I was convinced that the  "begun" state signifies any operation that has affected the state of an array job. Which in this case was deleting the subjobs in the Job array.

@agrawalravi90 , Job array goes to "begun" state only after the held job is released using "qrls". When an array Job is released from held, it's state is re-evaluated based on the sate of all the subjobs. It is then the job goes to Begun state, since the count of subjobs has changed from the time it was submitted. 

The "begun" state was not handled in the routing function, due to which the job was not getting routed from the routing queue. The fix I've provided has resolved this issue and hasn't changed any behavior, therefore I don't think there's any documentation required as part of this bug.

If we agree that begun state is not a valid state when a subjob is deleted, we can raise another bug/rfe to address this issue and handle it in separate ticket. The above fix has already been merged to pbspro-private and earlier 13.1.1 release.

**Comment by subhasisb** [comment]
Yew! Did I say that? That sounds so wrong! Do you remember why we thought that way, if we thought?

**Comment by subhasisb** [comment]
After discussing with Dilip, I think we have this understanding that:
a) We should not consider the state to be "begin" for any other operations on the job-array than atleast one sub-job starts execution. Ie, a subjob being deleted should not move the state to begin. This sounds like a different bug and should be filed and fixed separately.

b) Now once a job array has begun, it will always have the state begun till it is finished. You could qmove the job to a routing queue and still face this issue, so this bug fix approach still sounds valid.

**Comment by agrawalravi90** [comment]
Thanks for the clarification Subhasis. 1) makes perfect sense, did you guys already file a new ticket? Should I file one?

For 2), just wondering, after an arrayjob has already begun, can it still be moved? The admin guide says:
 "Note that job arrays can only be moved from one server to another if they are in the ‘Q’, ‘H’, or ‘W’ states, and only if there are no running subjobs. The state of the job array is preserved, and the job array will run to completion on the new server."

So, will we ever come across a situation where an arrayjob is in state begun, and it got moved to a routing queue? If not, then I think we should not add a case for begun in job_route() and let it fall under the default case to display the error message for invalid state.

**Comment by agrawalravi90** [comment]
ok, i just realized that the admin guide section i quoted only talks about moving it from one server to another, not about one queue to another. I couldn't find anything on moving between queues for arrayjobs, can a job be moved between queues if it is in begun state?

**Comment by dilip-krishnan** [comment]
Hi Ravi,
  I haven't tried it, but I guess if all subjobs are in queue state after been re-queued , the parent job will be still in begun state and it can be moved from queue to queue

**Comment by agrawalravi90** [comment]
Thanks for your response Dilip. I'm not sure what the supported behavior is, but looking at the code inside req_movejob(), it actually seems like we will reject a qmove if the job is not in 'Q', 'W' or 'H' state, regardless of whether it is a local move or remote (that fork is made later inside svr_movejob()):

	if (jobp->ji_qs.ji_state != JOB_STATE_QUEUED &&
		jobp->ji_qs.ji_state != JOB_STATE_HELD &&
		jobp->ji_qs.ji_state != JOB_STATE_WAITING) {
#ifndef NDEBUG
		(void)sprintf(log_buffer, "(%s) %s, state=%d",
			__func__, msg_badstate, jobp->ji_qs.ji_state);
		log_event(PBSEVENT_DEBUG, PBS_EVENTCLASS_JOB, LOG_DEBUG,
			jobp->ji_qs.ji_jobid, log_buffer);
#endif /* NDEBUG */
		req_reject(PBSE_BADSTATE, 0, req);
		return;
	}

So, could you please test your hypothesis out? If one is not allowed to qmove an arrayjob in the Begun state, then I'm not sure about adding a valid case for the begun state inside job_route().

**Comment by dilip-krishnan** [comment]
Hi Ravi,
  I tried the scenario and the array job didn't move to another queue for the same reason you pointed out in the source.
But moving jobs from one execution queue to another doesn't go through job_route(). The begin case is required in job_route since the arrayJob state is not evaluated based on the state of all the subjobs but rather on count(which we already agreed to be another bug). Having this case will allow the array Job which are in begun state to move further.

**Comment by agrawalravi90** [comment]
Thanks for testing that out Dilip

I'm afraid I don't agree with this bugfix. We are not fixing the actual bug with this code change, in fact we are making it worse, we are condoning it. The actual bug happens because of the fact that an arrayjob's state changes from H to B when a subjob is deleted. It would have gotten routed just fine if it were still in the H state when it hit job_route(). An arrayjob should never be in the B state when it is in a routing queue, job_route() should treat that as an invalid state for that operation and handle it accordingly, which it does already, so we should not change that.

So, to be consistent with our documentation, I recommend that we fix the code to make sure that an arrayjob doesn't go to state B unless a subjob has started (the function svr_evaljobstate()), and leave job_route() as it is. This bug will go away if we just make sure that the arrayjob is not transitioning to state B erroneously. What do you think?

**Comment by dilip-krishnan** [comment]
Hi Ravi,
    I had the same feeling about this bug and fix, it was an urgent bug fix for an earlier release and after discussing with @subhasisb, we realized that the repercussion of the correct fix would be higher and verification might be time consuming.  

But now we have time so we can definitely go ahead with correct fix instead. Thanks for sticking to the right thing to do here. I will close this pull request for now and modify the ticket tittle and description for PP-315, and will raise another pull request with correct fix.

Thanks again

**Pull request closed without merging**